toolkit/components/uniffi-bindgen-gecko-js/src/interface_ext.rs

```diff
@@ -222,3 +226,118 @@ pub impl FFIArgument {
         matches!(self.type_(), FFIType::RustBuffer)
     }
 }
+
+
+#[ext]
+pub impl Record {
+    fn js_name(&self) -> String {
+        self.name().to_camel_case()
+    }
+
+    fn js_constructor_field_list(&self) -> String {
+        let mut field_list = String::new();
+        for (idx, field) in self.fields().iter().enumerate() {
+            let field = field.name().to_mixed_case();
+            field_list.push_str(&field);
+            if idx != self.fields().len() - 1 {
+                field_list.push_str(",")
+            }
+        }
+        field_list
+    }
+
+    fn js_var_name(&self) -> String {
+        self.name().to_mixed_case()
+    }
+}
+
+#[ext]
+pub impl Field {
+    fn js_name(&self) -> String {
+        self.name().to_mixed_case()
+    }
+
+    fn write_datastream_fn(&self) -> String {
+        self.type_().write_datastream_fn()
+    }
+
+    fn read_datastream_fn(&self) -> String {
+        self.type_().read_datastream_fn()
+    }
+
+    fn js_ffi_converter(&self) -> String {
+        self.type_().js_ffi_converter()
+    }
+}
+
+#[ext]
+pub impl Argument {
+    fn js_lift_fn_name(&self) -> String {
+        format!("FfiConverter{}", self.type_().js_name())
+    }
+
+    fn js_name(&self) -> String {
+        self.name().to_mixed_case()
+    }
+}
+
+#[ext]
+pub impl Type {
+    fn js_name(&self) -> String {
+        match self {
+            Type::Float64 => "Double".to_string(),
+            Type::Optional(inner) => format!("Optional{}", inner.js_name()),
+            Type::Record(name) => name.to_camel_case(),
+            _ => todo!()
+        }
+    }
+
+    fn write_datastream_fn(&self) -> String {
+        match self {
+            Type::Float64 => format!("FfiConverterDouble.write"),
+            Type::Record(name) => format!("FfiConverter{}.write", name.to_camel_case()),
+            Type::Optional(inner) => format!("FfiConverterOptional{}.write", inner.js_name()),
+            _ => todo!()
+        }
+    }
+
+    fn read_datastream_fn(&self) -> String {
+        match self {
+            Type::Float64 => format!("FfiConverterDouble.read"),
+            Type::Record(name) => format!("FfiConverter{}.read", name.to_camel_case()),
+            Type::Optional(inner) => format!("FfiConverterOptional{}.read", inner.js_name()),
+            _ => todo!()
+        }
+    }
+
+    fn js_ffi_converter(&self) -> String {
+        match self {
+            Type::Float64 => "FfiConverterDouble".to_string(),
+            Type::Record(name) => format!("FfiConverter{}", name.to_camel_case()),
+            Type::Optional(inner) => format!("FfiConverterOptional{}", inner.js_name()),
+            _ => todo!()
+        }
+    }
```

Comment:
These `match` expressions might get out of control as we implement more types.  Two suggestions here:

- For the ffi-converter ones, I think you can leverage `canonical_name()` to avoid matching:
   - `js_ffi_converter(&self)` -> `format!("FfiConverter{}", self.canonical_name())`
   - `read_datastream_fn(&self)` -> `format!("{}.read", self.js_ffi_converter())`
   - ...
- If the only method that needs a match is `js_name()` then I think we're good.  If we end up adding several more methods that match in the same way, then we could consider using the pattern where we map a type to a `CodeType`, then which contains the implementation for those methods.

Reply:
Ya, I was a little lazy there 😅 Totally agree 💯